enterprise/frontend/src/metabase-enterprise/data_editing/tables/edit/use-table-crud.tsx

```diff
@@ -63,48 +78,131 @@ export const useTableCRUD = ({
   }, [tableMetadata]);
 
   const displayErrorIfExists = useCallback(
-    (error: any) => {
+    (error: unknown) => {
       if (error) {
         dispatch(
           addUndo({
             icon: "warning",
             toastColor: "error",
-            message: error?.data?.errors?.[0].error ?? t`An error occurred`,
+            message: getResponseErrorMessage(error) ?? t`An error occurred`,
           }),
         );
       }
     },
     [dispatch],
   );
 
+  const handleCellValueUpdateError = useCallback(
+    (
+      error: unknown,
+      cellUpdateContext: {
+        cellId: DataGridCellId;
+        patchResult: UndoObject | undefined;
+      },
+    ) => {
+      const { cellId, patchResult } = cellUpdateContext;
+
+      patchResult?.undo();
+
+      dispatch(
+        addUndo({
+          toastColor: "bg-black",
+          icon: null,
+          renderChildren: () => <ErrorUpdateToast error={error} />,
+          timeout: null, // removes automatic toast hide
+          undo: false,
+          onDismiss: () => {
+            // TODO: handle case when there are 2+ failed updates for a single cell id
+            setCellsWithFailedUpdatesMap((prevState) => {
+              const newMap = { ...prevState };
+              delete newMap[cellId];
+              return newMap;
+            });
+          },
+
+          // TODO: add list of open failed updates toast ids, clean them in case of filters update
+        }),
+      );
+
+      setCellsWithFailedUpdatesMap({
+        ...cellsWithFailedUpdatesMap,
+        [cellId]: true,
+      });
+    },
+    [cellsWithFailedUpdatesMap, dispatch],
+  );
+
   const handleCellValueUpdate = useCallback(
-    async ({ updatedData, rowIndex }: UpdatedRowCellsHandlerParams) => {
+    async ({ updatedData, rowIndex, cellId }: UpdateCellValueHandlerParams) => {
+      // mostly the same as "handleRowUpdate", but has optimistic update and special error handling
       if (!datasetData) {
         console.warn(
           "Failed to update table data - no data is loaded for a table",
         );
         return;
       }
 
-      const columns = datasetData.cols;
-      const rowData = datasetData.rows[rowIndex];
+      const pkRecord = getRowPkKeyValue(datasetData, rowIndex);
+      const updatedRowWithPk = {
+        ...updatedData,
+        ...pkRecord,
+      };
 
-      const pkColumnIndex = columns.findIndex(isPK);
-      const pkColumn = columns[pkColumnIndex];
-      const rowPkValue = rowData[pkColumnIndex];
+      const patchResult = stateUpdateStrategy.onRowsUpdated([updatedRowWithPk]);
 
+      try {
+        const response = await updateTableRows({
+          tableId: tableId,
+          rows: [updatedRowWithPk],
+        });
+
+        if (response.error) {
```

Comment:
I think we should apply state changes from an actual API response, just to be sure that the data is 100% parity with BE, so:
* Perform optimistic update
* Wait for response
* If response is successful, run state update strategy again with BE values